### PR TITLE
Introduce optional arguments in ccpp schemes and metadata

### PR DESCRIPTION
**Pull request by climbfuji**
## Description

This PR makes necessary changes for supporting optional arguments in CCPP (see https://github.com/ufs-community/ufs-weather-model/pull/2205 and PRs listed there). This involves adding the `optional` attribute tot he CCPP metadata for all variables that have matching host model variable with an active attribute other than the default value `.true.` (i.e. it may not be allocated). In addition, each of these variables need to be declared as `optional` variable in the Fortran code.

90% of all the changes in this PR were contributed by @dustinswales - kudos to him. The new and clean way to deal with optional arguments also allowed us to discover a few bugs/inconsistencies that we had to fix to get b4b identical results (and successful runs in the first place) with Intel and GNU on Hera.

Two source files need to be compiled with slightly lower optimization (`-O1` instead of `-O2` with Intel in release mode) to work around what seems to be bugs in the Intel compiler: `gcycle.F90` and `mynnedmf_wrapper.F90`). 
This should have negligible effect on the runtime, and it does not affect the results (see below).

This is part of a large set of PRs:

https://github.com/NCAR/ccpp-framework/pull/552 
https://github.com/NOAA-EMC/fv3atm/pull/807 
[https://github.com/ufs-community/ufs-weather-model/pull/2205](https://github.com/ufs-community/ufs-weather-model/pull/2205/files)
https://github.com/ufs-community/ccpp-physics/pull/189 
https://github.com/NCAR/ccpp-framework/pull/556 (can be scheduled and merged anytime beforehand)
https://github.com/NOAA-GFDL/GFDL_atmos_cubed_sphere/pull/338 
https://github.com/NOAA-PSL/stochastic_physics/pull/79

### Issue(s) addressed

Working towards https://github.com/NCAR/ccpp-framework/issues/540

## Testing

ufs-weather-model full regression testing on Hera (all tests b4b)

### CODE REVIEW

**Comment by climbfuji** [comment]
This PR is ready for review!

**Comment by dustinswales** [comment]
> This PR is ready for review!

@climbfuji Excellent work bringing this to the finish line!

@grantfirl @Qingfu-Liu I wonder why this PR didn't trigger review requests from A LOT of codeowners?
Turns out I didn't update [the CODEOWNERS file](https://github.com/ufs-community/ccpp-physics/blob/ufs/dev/CODEOWNERS) with the repo reorg. I will fix this now.

@climbfuji Can I open a PR into this branch updating the CODEOWNERS file?

**Comment by grantfirl** [comment]
> > This PR is ready for review!
> 
> @climbfuji Excellent work bringing this to the finish line!
> 
> @grantfirl @Qingfu-Liu I wonder why this PR didn't trigger review requests from A LOT of codeowners? Turns out I didn't update [the CODEOWNERS file](https://github.com/ufs-community/ccpp-physics/blob/ufs/dev/CODEOWNERS) with the repo reorg. I will fix this now.
> 
> @climbfuji Can I open a PR into this branch updating the CODEOWNERS file?

@dustinswales Ya, I just noticed this too. At first I thought that it was because there was one code owner who wasn't added as a collaborator with write permissions throwing off the process, but then I noticed that all of the paths were no longer valid! I think that it makes more sense to change the CODEOWNERS file in the next PR to be tested, which should be #205 .

**Comment by dustinswales** [comment]
> > > This PR is ready for review!
> > 
> > 
> > @climbfuji Excellent work bringing this to the finish line!
> > @grantfirl @Qingfu-Liu I wonder why this PR didn't trigger review requests from A LOT of codeowners? Turns out I didn't update [the CODEOWNERS file](https://github.com/ufs-community/ccpp-physics/blob/ufs/dev/CODEOWNERS) with the repo reorg. I will fix this now.
> > @climbfuji Can I open a PR into this branch updating the CODEOWNERS file?
> 
> @dustinswales Ya, I just noticed this too. At first I thought that it was because there was one code owner who wasn't added as a collaborator with write permissions throwing off the process, but then I noticed that all of the paths were no longer valid! I think that it makes more sense to change the CODEOWNERS file in the next PR to be tested, which should be #205 .

@grantfirl Sounds good. I'll open a PR into #205 this morning.

**Comment by dustinswales** [comment]
@grantfirl @climbfuji I opened PR's updating the CODEOWNERS into this PR and #205

**Comment by grantfirl** [comment]
@climbfuji @dustinswales I'm assuming that there are appropriate errors from the framework to catch errors/mismatches (i.e. optional attribute missing in physics metadata, active attribute missing in host, etc.)? For future reference, is there any other help for debugging this feature when errors related to this functionality pop up? @mkavulich @dustinswales We should probably add something in https://github.com/NCAR/ccpp-doc related to this.

**Comment by grantfirl** [comment]
> @climbfuji @dustinswales I'm assuming that there are appropriate errors from the framework to catch errors/mismatches (i.e. optional attribute missing in physics metadata, active attribute missing in host, etc.)? For future reference, is there any other help for debugging this feature when errors related to this functionality pop up? @mkavulich @dustinswales We should probably add something in https://github.com/NCAR/ccpp-doc related to this.

@climbfuji @dustinswales To clarify, is this adding a burden on reviewers to doublecheck that host active attribute matches ccpp-physics optional attribute matches Fortran code? Your answer will actually be helpful in reviewing this one -- whether I need to go through carefully double-checking all of this, or whether the framework would catch these kinds of errors.

**Comment by grantfirl** [comment]
@dustinswales @climbfuji Is the `True` value case-sensitive in the metadata file? Can the framework handle `true` or `T` or `t`? Is there any reason for someone to put `optional = False`?

**Comment by dustinswales** [comment]
> @climbfuji @dustinswales I'm assuming that there are appropriate errors from the framework to catch errors/mismatches (i.e. optional attribute missing in physics metadata, active attribute missing in host, etc.)? For future reference, is there any other help for debugging this feature when errors related to this functionality pop up? @mkavulich @dustinswales We should probably add something in https://github.com/NCAR/ccpp-doc related to this.

@grantfirl Yes, there are compatibility checks in place to ensure consistency between the host and the physics. The logic is described in [ccpp-framework Issue #526](https://github.com/NCAR/ccpp-framework/issues/526)

**Comment by dustinswales** [comment]
> > @climbfuji @dustinswales I'm assuming that there are appropriate errors from the framework to catch errors/mismatches (i.e. optional attribute missing in physics metadata, active attribute missing in host, etc.)? For future reference, is there any other help for debugging this feature when errors related to this functionality pop up? @mkavulich @dustinswales We should probably add something in https://github.com/NCAR/ccpp-doc related to this.
> 
> @climbfuji @dustinswales To clarify, is this adding a burden on reviewers to doublecheck that host active attribute matches ccpp-physics optional attribute matches Fortran code? Your answer will actually be helpful in reviewing this one -- whether I need to go through carefully double-checking all of this, or whether the framework would catch these kinds of errors.

@grantfirl When doing this we relied on Capgens parser, which parses the metadata and fortran source files for consistency. [As part of this effort, the parser was expanded to handle optional arguments](https://github.com/NCAR/ccpp-framework/pull/558).
So no you do not need to double check the meta/F90 files for consistency.

**Comment by dustinswales** [comment]
> @dustinswales @climbfuji Is the `True` value case-sensitive in the metadata file? Can the framework handle `true` or `T` or `t`? Is there any reason for someone to put `optional = False`?

I don't believe it's case sensitive, but I'm pretty certain we don't allow for aliases in the metadata. @peverwhee Do you recall?

**Comment by peverwhee** [comment]
> > @dustinswales @climbfuji Is the `True` value case-sensitive in the metadata file? Can the framework handle `true` or `T` or `t`? Is there any reason for someone to put `optional = False`?
> 
> I don't believe it's case sensitive, but I'm pretty certain we don't allow for aliases in the metadata. @peverwhee Do you recall?

@dustinswales yep. It's not case-sensitive. You can also alias as "T" and "F" (or "t" and "f") and can use fortran syntax (.true. and .false.)

**Comment by dustinswales** [comment]
> > > @dustinswales @climbfuji Is the `True` value case-sensitive in the metadata file? Can the framework handle `true` or `T` or `t`? Is there any reason for someone to put `optional = False`?
> > 
> > 
> > I don't believe it's case sensitive, but I'm pretty certain we don't allow for aliases in the metadata. @peverwhee Do you recall?
> 
> @dustinswales yep. It's not case-sensitive. You can also alias as "T" and "F" (or "t" and "f") and can use fortran syntax (.true. and .false.)

Excellent. Thanks!

**Review by grantfirl — APPROVED** [review]
With so many files changed, thanks for keeping the changes "targeted" to the task at hand.

**Review by JongilHan66 — APPROVED** [review]

**Review by mdtoyNOAA — APPROVED** [review]

File: physics/MP/NSSL/mp_nssl.F90 [hunk 1]
```diff
@@ -68,12 +68,12 @@ subroutine mp_nssl_init(ncol, nlev, errflg, errmsg, threads, restart, &
          real(kind_phys),           intent(inout) :: qi (:,:) !(1:ncol,1:nlev)
          real(kind_phys),           intent(inout) :: qs (:,:) !(1:ncol,1:nlev)
          real(kind_phys),           intent(inout) :: qh (:,:) !(1:ncol,1:nlev) graupel
-         real(kind_phys),           intent(inout) :: ccw(:,:) !(1:ncol,1:nlev)
+         real(kind_phys),           intent(inout), optional :: ccw(:,:) !(1:ncol,1:nlev)
```
Comment (MicroTed) [inline physics/MP/NSSL/mp_nssl.F90:71]: Note: ccw should always be active (i.e., not optional)

File: physics/SFC_Models/Land/RUC/lsm_ruc.meta [hunk 2]
```diff
@@ -320,6 +320,7 @@
   type = real
   kind = kind_phys
   intent = inout
+  optional = True
```
Comment (tanyasmirnova) [inline physics/SFC_Models/Land/RUC/lsm_ruc.meta:323]: @climbfuji @dustinswales The majority of variables that you add optional to, are not optional. For example smois - it is a state soil moisture variable.

**Review by barlage — APPROVED** [review]

**Comment by dustinswales** [comment]
@tanyasmirnova @MicroTed 
Any host field that was conditionally allocated, [for example smois](https://github.com/NOAA-EMC/fv3atm/blob/develop/ccpp/data/GFS_typedefs.meta#L1812), now has the optional attribute defined in the physics.
This does not mean that these fields should be optional, but rather a relic of whatever logic was controlling these fields on the host side.

**Review by tanyasmirnova — APPROVED** [review]

**Review by AnningCheng-NOAA — APPROVED** [review]
look fine with me.

**Comment by MicroTed** [comment]
@dustinswales 
> @tanyasmirnova @MicroTed Any host field that was conditionally allocated ... This does not mean that these fields should be optional, but rather a relic of whatever logic was controlling these fields on the host side.

I'm not sure I understand why the CCPP interface should care about the host model this way? What if different host models have different logic? I do like that the optional attribute is back after being removed.

**Comment by FernandoAndrade-NOAA** [comment]
Testing for [#2205](https://github.com/ufs-community/ufs-weather-model/pull/2205) has completed successfully, please continue with merging this PR.

**Pull request merged**